sdk/azidentity/chained_token_credential.go

```diff
@@ -18,37 +19,39 @@ type ChainedTokenCredential struct {
 // NewChainedTokenCredential creates an instance of ChainedTokenCredential with the specified TokenCredential sources.
 func NewChainedTokenCredential(sources ...azcore.TokenCredential) (*ChainedTokenCredential, error) {
 	if len(sources) == 0 {
-		return nil, &CredentialUnavailableError{Message: "NewChainedTokenCredential: length of sources cannot be 0"}
+		return nil, &CredentialUnavailableError{CredentialType: "Chained Token Credential", Message: "Length of sources cannot be 0"}
 	}
-	// TODO: test for each of these conditions
+
 	for _, source := range sources {
 		if source == nil {
-			return nil, &CredentialUnavailableError{Message: "NewChainedTokenCredential: sources cannot contain a nil TokenCredential"}
+			return nil, &CredentialUnavailableError{CredentialType: "Chained Token Credential", Message: "Sources cannot contain a nil TokenCredential"}
 		}
 	}
 
 	return &ChainedTokenCredential{sources: sources}, nil
 }
 
 // GetToken sequentially calls TokenCredential.GetToken on all the specified sources, returning the first non default AccessToken.
-func (c *ChainedTokenCredential) GetToken(ctx context.Context, scopes []string) (*azcore.AccessToken, error) {
-	// CP: Notes from session with Jeff, have two funcs for the DefaultAzureCredential the main default func will issue the call that hits the wire, however internally first we create the chain and return a credUnavailable error if no credentials are found, if one works then we immediately return the one that works. End goal to always try to use the same credential type so we dont switch between credentials during execution, or at least reduce the possibility of that happening?
+func (c *ChainedTokenCredential) GetToken(ctx context.Context, opts *azcore.TokenRequestOptions) (*azcore.AccessToken, error) {
+	var credErr *CredentialUnavailableError
 	var token *azcore.AccessToken
 	var err error
-	var errList []error
-
-	for i := 0; i < len(c.sources) && token == nil; i++ {
-		token, err = c.sources[i].GetToken(ctx, scopes)
-		// CP: This currently works the same way as the other languages
-		// TODO: instead of appending to an error slice, check if the error returned is an auth failure then stop the function completely and return the auth failure, credunavailable is still ok to continue looping
-		if err != nil {
-			errList = append(errList, err)
+	var errList []*CredentialUnavailableError
+
+	for i := 0; i < len(c.sources); i++ {
+		token, err = c.sources[i].GetToken(ctx, opts)
+		if errors.As(err, &credErr) {
+			errList = append(errList, err.(*CredentialUnavailableError))
+		} else {
+			if token != nil && err == nil {
```

Comment:
update logic, use else if and return token outside of if block

Reply:
else if err != nil { nil, err}

return token, nil

Reply:
Done, except I put the return token, nil in the else statement. Feel free to comment again if you disagree.